File: kangtaku/valid_palindrome_II.cpp
class Solution {
public:
    bool validPalindrome(string s) {
        int l = 0, r = s.length() - 1;
        while (l < r) {
            if (s[l] != s[r]) {
                return isPalindrome(s, l + 1, r) || isPalindrome(s, l, r - 1);
            }
            l++;
            r--;
        }
        return true;
    }
    
    bool isPalindrome(string s, int l, int r) {
        while (l < r) {
            if (s[l] != s[r]) {
                return false;
            }
            l++;
            r--;
        }
        return true;
    }
};

File: byssup/string/Valid Palindrome II.py
class Solution(object):
    def validPalindrome(self, s):
        """
        :type s: str
        :rtype: bool
        """
        i, j = self.get_idx(s)
        if i > j:
            return True
        ii, jj = self.get_idx(s[:i] + s[i+1:])
        if ii > jj:
            return True
        ii, jj = self.get_idx(s[:j] + s[j+1:])
        if ii > jj:
            return True
        return False

    def get_idx(self, s):
        i, j = 0, len(s) - 1
        while i <= j:
            if s[i] != s[j]:
                break
            i += 1
            j -= 1
        return i, j